pkg/api/service.go

```diff
@@ -55,15 +64,63 @@ func (s *Service) Close() {
 
 func (s *Service) BatchSubscribeEnvelopes(
 	req *message_api.BatchSubscribeEnvelopesRequest,
-	server message_api.ReplicationApi_BatchSubscribeEnvelopesServer,
+	stream message_api.ReplicationApi_BatchSubscribeEnvelopesServer,
 ) error {
-	return status.Errorf(codes.Unimplemented, "method BatchSubscribeEnvelopes not implemented")
+	log := s.log.Named("subscribe")
+
+	// Send a header (any header) to fix an issue with Tonic based GRPC clients.
+	// See: https://github.com/xmtp/libxmtp/pull/58
+	err := stream.SendHeader(metadata.Pairs("subscribed", "true"))
+	if err != nil {
+		return status.Errorf(codes.Internal, "could not send header: %v", err)
+	}
+
+	requests := req.GetRequests()
+	if len(requests) == 0 {
+		return status.Errorf(codes.InvalidArgument, "missing requests")
+	}
+
+	ch, err := s.subscribeWorker.addListeners(requests)
+	if err != nil {
+		return status.Errorf(codes.InvalidArgument, "invalid subscription request: %v", err)
+	}
+
+	var streamLock sync.Mutex
```

Comment:
what is this lock protecting? Who is racing against who here?

Reply:
Great question. The loop below actually came from the implementation on the old server. The lock is used for the duration of the send operation on the stream, but I actually don't see any need for it because the loop is running synchronously. Maybe it was some kind of historical cruft - I'll remove it tomorrow as well as make the style changes we're requesting.